Update for version 0.14.0

With version 0.14.0 this extension is not able to be instantiated anymore. 
My guess is that the constructor should be adjusted so that the prefix is not needed anymore.

```code
[ERROR] [11/04/2016 12:30:56.936] [my-app-akka.actor.default-dispatcher-18] [akka://my-app-akka/user/$a] A Coordination implementation must provide a constructor with the following signature:
(clusterName: String, system: ActorSystem)
akka.actor.ActorInitializationException: akka://my-app/user/$a: exception during creation
	at akka.actor.ActorInitializationException$.apply(Actor.scala:174)
	at akka.actor.ActorCell.create(ActorCell.scala:607)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:461)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
	at akka.dispatch.Mailbox.run(Mailbox.scala:223)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.Exception: A Coordination implementation must provide a constructor with the following signature:
(clusterName: String, system: ActorSystem)
	at de.heikoseeberger.constructr.coordination.Coordination$.apply(Coordination.scala:37)
	at de.heikoseeberger.constructr.Constructr.createConstructrMachine(Constructr.scala:91)
	at de.heikoseeberger.constructr.Constructr.<init>(Constructr.scala:61)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at java.lang.Class.newInstance(Class.java:442)
	at akka.util.Reflect$.instantiate(Reflect.scala:44)
	at akka.actor.NoArgsReflectConstructor.produce(IndirectActorProducer.scala:105)
	at akka.actor.Props.newActor(Props.scala:213)
	at akka.actor.ActorCell.newActor(ActorCell.scala:562)
	at akka.actor.ActorCell.create(ActorCell.scala:588)
	... 9 more
```

--- a/constructr-akka-testing/src/multi-jvm/scala/com/github/everpeace/constructor/coordination/redis/MultiNodeRedisConductrSpec.scala
+++ b/constructr-akka-testing/src/multi-jvm/scala/com/github/everpeace/constructor/coordination/redis/MultiNodeRedisConductrSpec.scala
@@ -24,6 +24,5 @@ class MultiNodeRedisConstructrSpecMultiJvmNode5 extends MultiNodeRedisConstructr
 
 abstract class MultiNodeRedisConstructrSpec extends MultiNodeRedisConstructrSpecBase(
   6379,
-  "akka",
   "MultiNodeRedisConstructrSpecBase"
 )

--- a/constructr-akka-testing/src/multi-jvm/scala/com/github/everpeace/constructor/coordination/redis/MultiNodeRedisConstructrSpecBase.scala
+++ b/constructr-akka-testing/src/multi-jvm/scala/com/github/everpeace/constructor/coordination/redis/MultiNodeRedisConstructrSpecBase.scala
@@ -16,8 +16,7 @@
 
 package com.github.everpeace.constructor.coordination.redis
 
-import akka.actor.ActorDSL.{Act, actor}
-import akka.actor.Address
+import akka.actor.{Actor, Props}
 import akka.cluster.{Cluster, ClusterEvent}
 import akka.pattern.ask
 import akka.remote.testkit.{MultiNodeConfig, MultiNodeSpec}
@@ -26,7 +25,7 @@ import akka.testkit.TestDuration
 import akka.util.Timeout
 import com.github.everpeace.constructr.coordination.redis.RedisClientFactory
 import com.typesafe.config.ConfigFactory
-import de.heikoseeberger.constructr.akka.ConstructrExtension
+import de.heikoseeberger.constructr.ConstructrExtension
 import de.heikoseeberger.constructr.coordination.Coordination
 import org.scalatest.{BeforeAndAfterAll, FreeSpecLike, Matchers}
 
@@ -60,13 +59,13 @@ class ConstructrMultiNodeConfig(coordinationPort: Int) extends MultiNodeConfig {
   }
 }
 
-abstract class MultiNodeRedisConstructrSpecBase(coordinationPort: Int, prefix: String, clusterName: String)
+abstract class MultiNodeRedisConstructrSpecBase(coordinationPort: Int, clusterName: String)
   extends MultiNodeSpec(new ConstructrMultiNodeConfig(coordinationPort))
     with FreeSpecLike with Matchers with BeforeAndAfterAll {
 
   implicit val mat = ActorMaterializer()
   val redis = RedisClientFactory.fromConfig(system.settings.config)
-  val redisCoordination = Coordination(prefix, clusterName, system)
+  val redisCoordination = Coordination(clusterName, system)
 
   "Constructr should manage an Akka cluster" in {
     runOn(roles.head) {
@@ -84,18 +83,17 @@ abstract class MultiNodeRedisConstructrSpecBase(coordinationPort: Int, prefix: S
     enterBarrier("coordination-started")
 
     ConstructrExtension(system)
-    val listener = actor(new Act {
-
+    val listener = system.actorOf(Props(new Actor {
       import ClusterEvent._
 
       var isMember = false
       Cluster(context.system).subscribe(self, InitialStateAsEvents, classOf[MemberJoined], classOf[MemberUp])
-      become {
+      def receive: Receive = {
         case "isMember" => sender() ! isMember
         case MemberJoined(member) if member.address == Cluster(context.system).selfAddress => isMember = true
         case MemberUp(member) if member.address == Cluster(context.system).selfAddress => isMember = true
       }
-    })
+    }))
     within(20.seconds.dilated) {
       awaitAssert {
         implicit val timeout = Timeout(1.second.dilated)
@@ -108,9 +106,8 @@ abstract class MultiNodeRedisConstructrSpecBase(coordinationPort: Int, prefix: S
 
     within(5.seconds.dilated) {
       awaitAssert {
-        import de.heikoseeberger.constructr.akka._
         val constructrNodes = Await.result(
-          redisCoordination.getNodes[Address](),
+          redisCoordination.getNodes(),
           1.second.dilated
         )
         roles.to[Set].map(_.name.toInt) shouldEqual constructrNodes.flatMap(_.port)

--- a/constructr-coordination-redis/src/main/scala/com/github/everpeace/constructr/coordination/redis/RedisCoordination.scala
+++ b/constructr-coordination-redis/src/main/scala/com/github/everpeace/constructr/coordination/redis/RedisCoordination.scala
@@ -17,48 +17,42 @@
 package com.github.everpeace.constructr.coordination
 package redis
 
-import java.util.Base64
-
 import _root_.redis.RedisClient
 import akka.Done
-import akka.actor.{ ActorRefFactory, ActorSystem }
-import com.typesafe.config.{ Config, ConfigException }
+import akka.actor.{ActorRefFactory, ActorSystem, Address, AddressFromURIString}
+import com.typesafe.config.{Config, ConfigException}
 import de.heikoseeberger.constructr.coordination.Coordination
-import de.heikoseeberger.constructr.coordination.Coordination.NodeSerialization
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
 import scala.concurrent.duration.FiniteDuration
 
 final class RedisCoordination(
-    val prefix: String,
     val clusterName: String,
     val system: ActorSystem
 ) extends Coordination {
-  implicit val _system: ActorRefFactory = system
-  implicit val _ec = _system.dispatcher
+  private[this] implicit val _system: ActorRefFactory = system
+  private[this] implicit val _ec: ExecutionContext = _system.dispatcher
 
-  val redis = RedisClientFactory.fromConfig(system.settings.config)
+  private[this] val redis = RedisClientFactory.fromConfig(system.settings.config)
 
-  def lockKey = s"${prefix}/${clusterName}/lock"
-  def nodeKeyPrefix = s"${prefix}/${clusterName}/nodes"
-  def nodeKey[A: NodeSerialization](self: A) = s"${nodeKeyPrefix}/${Base64.getUrlEncoder.encodeToString(NodeSerialization.toBytes(self))}"
+  def lockKey = s"$clusterName/lock"
+  def nodeKeyPrefix = s"$clusterName/nodes"
+  def nodeKey(self: Address) = s"$nodeKeyPrefix/${self.toString}"
 
   /**
    * Get the nodes.
-   *
-   * @tparam A node type, must have a [[Coordination.NodeSerialization]]
    * @return future of nodes
    */
-  override def getNodes[A: NodeSerialization](): Future[Set[A]] = redis.keys(s"${nodeKeyPrefix}/*").map(ks => scala.collection.immutable.Seq(ks: _*)).flatMap { ks =>
+  override def getNodes(): Future[Set[Address]] = redis.keys(s"$nodeKeyPrefix/*").map(ks => scala.collection.immutable.Seq(ks: _*)).flatMap { ks =>
     // N+1 problem would happen here.
     // But N might not be so large in real situation.
-    val f: Seq[Future[Option[A]]] = for {
+    val f: Seq[Future[Option[Address]]] = for {
       k <- ks
     } yield {
       for {
-        bytesOpt <- redis.get[Array[Byte]](k)
+        bytesOpt <- redis.get[String](k)
       } yield {
-        bytesOpt.map(NodeSerialization.fromBytes[A])
+        bytesOpt.map(AddressFromURIString(_))
       }
     }
     Future.sequence(f).map { opts =>
@@ -71,21 +65,20 @@ final class RedisCoordination(
   }
 
   /**
-   * Akquire a lock for bootstrapping the cluster (first node).
+   * Acquire a lock for bootstrapping the cluster (first node).
    *
    * @param self self node
    * @param ttl  TTL for the lock
-   * @tparam A node type, must have a [[Coordination.NodeSerialization]]
    * @return true, if lock could be akquired, else false
    */
-  override def lock[A: NodeSerialization](self: A, ttl: FiniteDuration): Future[Boolean] = {
-    def readLock(): Future[Option[A]] = redis.get[Array[Byte]](lockKey).map(_.map(NodeSerialization.fromBytes[A]))
-    def writeLock(lockHolder: A): Future[Boolean] = {
+  override def lock(self: Address, ttl: FiniteDuration): Future[Boolean] = {
+    def readLock(): Future[Option[Address]] = redis.get[String](lockKey).map(_.map(AddressFromURIString(_)))
+    def writeLock(lockHolder: Address): Future[Boolean] = {
       val trans2 = redis.watch(lockKey)
-      val writeLock = trans2.set(lockKey, NodeSerialization.toBytes(lockHolder), None, Some(ttl.toMillis))
+      val writeLock = trans2.set(lockKey, lockHolder.toString, None, Some(ttl.toMillis))
       trans2.exec()
       writeLock.recover {
-        case error => false
+        case _ => false
       }
     }
     readLock().flatMap {
@@ -99,26 +92,24 @@ final class RedisCoordination(
    *
    * @param self self node
    * @param ttl  TTL for the node entry
-   * @tparam A node type, must have a [[Coordination.NodeSerialization]]
    * @return future signaling done
    */
-  override def refresh[A: NodeSerialization](self: A, ttl: FiniteDuration): Future[Done] = addSelfOrRefresh(self, ttl)
+  override def refresh(self: Address, ttl: FiniteDuration): Future[Done] = addSelfOrRefresh(self, ttl)
 
   /**
    * Add self to the nodes.
    *
    * @param self self node
    * @param ttl  TTL for the node entry
-   * @tparam A node type, must have a [[Coordination.NodeSerialization]]
    * @return future signaling done
    */
-  override def addSelf[A: NodeSerialization](self: A, ttl: FiniteDuration): Future[Done] = addSelfOrRefresh(self, ttl)
+  override def addSelf(self: Address, ttl: FiniteDuration): Future[Done] = addSelfOrRefresh(self, ttl)
 
   // redis doesn't support element-wise ttl in [sorted] sets.
   // So, this sets with
   //   key -> {prefix}/{clusterName}/nodes/{encoded self}
   //   value -> encoded self
-  private def addSelfOrRefresh[A: NodeSerialization](self: A, ttl: FiniteDuration): Future[Done] = redis.set(nodeKey(self), NodeSerialization.toBytes(self), None, Some(ttl.toMillis)).map(_ => Done)
+  private def addSelfOrRefresh(self: Address, ttl: FiniteDuration): Future[Done] = redis.set(nodeKey(self), self.toString, None, Some(ttl.toMillis)).map(_ => Done)
 
 }
 
@@ -132,12 +123,12 @@ object RedisClientFactory {
     val password = try {
       Option(config.getString("constructr.coordination.redis.password")).filter(_.nonEmpty)
     } catch {
-      case e: ConfigException.Missing => None
+      case _: ConfigException.Missing => None
     }
     val db = try {
       Option(config.getInt("constructr.coordination.redis.db"))
     } catch {
-      case e: ConfigException.Missing => None
+      case _: ConfigException.Missing => None
     }
 
     RedisClient(host, port, password, db)

--- a/constructr-coordination-redis/src/test/scala/com/github/everpeace/constructr/coordination/redis/RedisCoordinationSpec.scala
+++ b/constructr-coordination-redis/src/test/scala/com/github/everpeace/constructr/coordination/redis/RedisCoordinationSpec.scala
@@ -16,27 +16,18 @@
 
 package com.github.everpeace.constructr.coordination.redis
 
-import java.nio.charset.StandardCharsets._
-
 import akka.Done
-import akka.actor.ActorSystem
+import akka.actor.{ ActorSystem, Address }
 import akka.testkit.TestProbe
 import com.typesafe.config.ConfigFactory
 import de.heikoseeberger.constructr.coordination.Coordination
 import org.scalatest.{ BeforeAndAfterAll, Matchers, WordSpec }
 
-import scala.concurrent.{ Awaitable, Await }
+import scala.concurrent.{ Await, Awaitable }
 import scala.concurrent.duration._
 import scala.util.Random
 
 object RedisCoordinationSpec {
-  import Coordination._
-
-  private implicit val stringNodeSerialization = new NodeSerialization[String] {
-    override def fromBytes(bytes: Array[Byte]) = new String(bytes, UTF_8)
-    override def toBytes(s: String) = s.getBytes(UTF_8)
-  }
-
   // this test assumes redis server is up on DOCKER_HOST or localhost(127.0.0.1)
   // below command would be help:
   //   $ docker run --name some-redis -p 6379:6379 -d redis
@@ -58,25 +49,28 @@ class RedisCoordinationSpec extends WordSpec with Matchers with BeforeAndAfterAl
 
   "RedisCoordination" should {
     "correctly interact with redis" in {
-      val coordination: Coordination = new RedisCoordination(randomString(), randomString(), system)
+      val coordination: Coordination = new RedisCoordination(randomString(), system)
 
-      resultOf(coordination.getNodes[String]()) shouldBe 'empty
+      resultOf(coordination.getNodes()) shouldBe 'empty
 
-      resultOf(coordination.lock[String]("self", 10.seconds)) shouldBe true
-      resultOf(coordination.lock[String]("self", 10.seconds)) shouldBe true
-      resultOf(coordination.lock[String]("other", 10.seconds)) shouldBe false
+      val selfAddress = Address("tcp", "self")
+      val otherAddress = Address("tcp", "other")
 
-      resultOf(coordination.addSelf[String]("self", 10.seconds)) shouldBe Done
-      resultOf(coordination.getNodes[String]()) shouldBe Set("self")
+      resultOf(coordination.lock(selfAddress, 10.seconds)) shouldBe true
+      resultOf(coordination.lock(selfAddress, 10.seconds)) shouldBe true
+      resultOf(coordination.lock(otherAddress, 10.seconds)) shouldBe false
 
-      resultOf(coordination.refresh[String]("self", 1.second)) shouldBe Done
-      resultOf(coordination.getNodes[String]()) shouldBe Set("self")
+      resultOf(coordination.addSelf(selfAddress, 10.seconds)) shouldBe Done
+      resultOf(coordination.getNodes()) shouldBe Set(selfAddress)
+
+      resultOf(coordination.refresh(selfAddress, 1.second)) shouldBe Done
+      resultOf(coordination.getNodes()) shouldBe Set(selfAddress)
 
       val probe = TestProbe()
       import probe._
       within(5.seconds) { // 2 seconds should be enough, but who knows hows ...
         awaitAssert {
-          resultOf(coordination.getNodes[String]()) shouldBe 'empty
+          resultOf(coordination.getNodes()) shouldBe 'empty
         }
       }
     }

--- a/project/Dependencies.scala
+++ b/project/Dependencies.scala
@@ -2,17 +2,17 @@ import sbt._
 
 object Version {
     final val Scala            = "2.11.8"
-    final val Constructr       = "0.13.2"
+    final val Constructr       = "0.17.0"
     final val Rediscala        = "1.6.0"
-    final val Akka             = "2.4.4"
+    final val Akka             = "2.5.0"
     final val AkkaLog4j        = "1.1.3"
     final val Log4j            = "2.5"
-    final val ScalaTest        = "2.2.6"
+    final val ScalaTest        = "3.0.0"
 }
 
 object Library {
   val constructrCoordination = "de.heikoseeberger"        %% "constructr-coordination"  % Version.Constructr
-  val constructrAkka         = "de.heikoseeberger"        %% "constructr-akka"          % Version.Constructr
+  val constructrAkka         = "de.heikoseeberger"        %% "constructr"               % Version.Constructr
   val rediscala              = "com.github.etaty"         %% "rediscala"                % Version.Rediscala
   val akkaCluster            = "com.typesafe.akka"        %% "akka-cluster"             % Version.Akka
   val akkaTestkit            = "com.typesafe.akka"        %% "akka-testkit"             % Version.Akka

